### PR TITLE
drivers: CAN: Abstract blocking send call

### LINKED ISSUE #22379: stm32: can_send gets stuck (ignoring timeout) if bus not present
If the transceiver is not wired to the bus, `can_stm32_send` gets stuck after 3 calls (= length of the internal TX queue of the used STM32F072) in the following line:
https://github.com/zephyrproject-rtos/zephyr/blob/ac12066800ac0d7c2e353462b830f96e8e97fe1b/drivers/can/can_stm32.c#L630

Changing the  `k_sem_take` timeout from `K_FOREVER` to `timeout` (= function argument) solved the problem. But I'm not sure if this is the ideal solution, @alexanderwachter .

I'm currently using the CAN interface in a very simple way without any callbacks:
`can_send(can_dev, &msg, K_MSEC(10), NULL, NULL);`

### PR DESCRIPTION
**Pull request by alexanderwachter**
This commit moves the semaphore for a blocking call from
the driver implementation to the stack of the caller.
It reduces the complexity of the drivers and saves some RAM.

Fixes #22379

### CODE REVIEW

File: include/drivers/can.h [hunk 1]
```diff
@@ -355,7 +363,22 @@ static inline int z_impl_can_send(struct device *dev,
...
 		(const struct can_driver_api *)dev->driver_api;
 
-	return api->send(dev, msg, timeout, callback_isr, callback_arg);
+	if (callback_isr) {
+		return api->send(dev, msg, timeout, callback_isr, callback_arg);
+	} else {
+		struct can_send_wait send_wait;
+		int ret;
+
+		k_sem_init(&send_wait.sem, 0, 1);
```
Comment (henrikbrixandersen) [inline include/drivers/can.h]: This change will introduce an overhead of initialising a semaphore on each `can_send()` call without any callback defined. I am not sure it is worth it.

It also seem very unlike any other API implementation in Zephyr?

Comment (karstenkoenig) [reply include/drivers/can.h]: Yea I am sort of ambivalent about it. Not sure how much of a trade-off the memory saved vs the calling overhead it is. 
In terms of driver complexity I think it isn't a big improvement, we all had the same approach and it's not so hard to follow either.

Comment (alexanderwachter) [reply include/drivers/can.h]: >This change will introduce an overhead of initialising a semaphore on each can_send() call without any callback defined. I am not sure it is worth it.

The overhead of initializing is not that much. The blocking send call is used rarely, and in the current implementation, we need to maintain a sem for every mailbox just in case we will need it.

>In terms of driver complexity I think it isn't a big improvement, we all had the same approach and it's not so hard to follow either.

The complexity raised with the timeout patch :-)
I really don't like #19707 in its current state.

Comment (karstenkoenig) [reply include/drivers/can.h]: Ah gotcha. What would be your plan then, use that can_common_isr_callback_handler() for failing sends with timeouts?

Comment (alexanderwachter) [reply include/drivers/can.h]: Updated #19707 on top of this PR. Looks better now!
